application/email/v0/read_emails.go

```diff
@@ -0,0 +1,274 @@
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer client.Close()
+
+	setup := e.GetSetup()
+	err = client.Login(
+		setup.GetFields()["email-address"].GetStringValue(),
+		setup.GetFields()["password"].GetStringValue(),
+	).Wait()
+	if err != nil {
+		return nil, err
+	}
+
+	emails, err := fetchEmails(client, inputStruct.Search)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := client.Logout().Wait(); err != nil {
+		return nil, err
+	}
+
+	outputStruct := ReadEmailsOutput{
+		Emails: emails,
+	}
+
+	output, err := base.ConvertToStructpb(outputStruct)
+	if err != nil {
+		return nil, err
+	}
+
+	return output, nil
+}
+
+func initIMAPClient(serverAddress string, serverPort int) (*imapclient.Client, error) {
+
+	c, err := imapclient.DialTLS(serverAddress+":"+strconv.Itoa(serverPort), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func fetchEmails(c *imapclient.Client, search Search) ([]Email, error) {
+
+	mailbox := search.Mailbox
+	imapMailbox := convertMailboxString(mailbox)
+
+	selectedMbox, err := c.Select(imapMailbox, nil).Wait()
+	if err != nil {
+		return nil, err
+	}
+
+	emails := []Email{}
+
+	if selectedMbox.NumMessages == 0 {
+		return emails, nil
+	}
+
+	if search.Limit == 0 {
+		search.Limit = EmailReadingDefaultCapacity
+	}
+	limit := search.Limit
+
+	// TODO: chuang8511, Research how to fetch emails by filter and concurrency.
+	// It will be done before 2024-07-26.
+	for i := selectedMbox.NumMessages; limit > 0; i-- {
+		limit--
+		email := Email{}
+		var seqSet imap.SeqSet
+		seqSet.AddNum(i)
+
+		fetchOptions := &imap.FetchOptions{
+			BodySection: []*imap.FetchItemBodySection{{}},
+		}
+		fetchCmd := c.Fetch(seqSet, fetchOptions)
+		msg := fetchCmd.Next()
+		var bodySection imapclient.FetchItemDataBodySection
+		var ok bool
+		for {
+			item := msg.Next()
+			if item == nil {
+				break
+			}
+			bodySection, ok = item.(imapclient.FetchItemDataBodySection)
+			if ok {
+				break
+			}
+		}
+		if !ok {
+			return nil, fmt.Errorf("FETCH command did not return body section")
+		}
+
+		mr, err := mail.CreateReader(bodySection.Literal)
+		if err != nil {
+			return nil, fmt.Errorf("FETCH command did not return body section")
+		}
+
+		h := mr.Header
+		setEnvelope(&email, h)
+
+		if !includeSearchCondition(email, search) {
+			if err := fetchCmd.Close(); err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		for {
+			p, err := mr.NextPart()
+			if err == io.EOF {
+				break
+			}
+			inlineHeader, ok := p.Header.(*mail.InlineHeader)
+			if ok && !isHTMLType(inlineHeader) {
+				b, _ := io.ReadAll(p.Body)
+				email.Message += string(b)
+			}
+		}
+
+		if !includeSearchMessage(email, search) {
+			if err := fetchCmd.Close(); err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		if err := fetchCmd.Close(); err != nil {
+			return nil, err
+		}
+
+		emails = append(emails, email)
+	}
+
+	return emails, nil
+}
+
+func convertMailboxString(mailbox string) string {
```

Comment:
Does this mean that the current implementation can only be used with Gmail?

Reply:
@donch1989 
I miss to consider this point when I developed it. Thanks for pointing it out.
Now, I have 2 ideas for this.

1. do not support mailbox field, and we fetch all emails from all mailbox.
2. make mailbox as plain text, and we write the Gmail input as the example in Email Component Document

What do you think about this?

Reply:
I think we can go with option 1 first. Will that have any side effects?

Reply:
> Will that have any side effects?

It will take a longer time to scan all mailboxes.
For AI SDR demo, I think the speed of fetching emails & accuracy of the content would be the key to success.
By specifying the mailbox, it can increase the performance in terms of speed and accuracy.

Reply:
Could we go with option 2 in the current phase?
Is only searching one mailbox a damage? If so, how about we set it as array:string?
Or, do you have other concerns that I miss?

Reply:
> It will take a longer time to scan all mailboxes.

I'm not sure about the mechanism here. I remember you already set a limit on the number of emails, right?

> array:string
I don't think array: string can work well on Console for now.

The problem with Option 2 is that users may struggle to type the correct mailbox name.

Reply:
> I'm not sure about the mechanism here. I remember you already set a limit on the number of emails, right?

Yes, the limit I set is the number to "scan", it will only scan a specific mailbox if the limit is smaller than the message number in a mailbox.
By the current design, it won't get the content for AI SDR usage if we make it scan INBOX first.

The purpose I specially design mailbox field is for the usage to get the sending mails specifically.

> The problem with Option 2 is that users may struggle to type the correct mailbox name.

Yes, it is the purpose to design this field with enum.

Reply:
> the limit I set is the number to "scan",

The reason of this design is that I have not found how to fetch emails through batch way.
So, we have to fetch email [one by one ](https://github.com/instill-ai/component/pull/178/files#diff-0382a2917f426e2ffd3a03998b35a9198571869579d4806de3a03eec10b16caeR129).

It makes the scanning takes time if we do not have limit here.
So, I added this field and set it 100 as default.

Reply:
OK, I just check the imap package, it need the mailbox input. Then I think we can use Option 2 with a string plaintext.

Reply:
Thank you! I will add Gmail mailbox in the document as well.